### PR TITLE
Feature/multi method per instrument

### PR DESCRIPTION
**Pull request by xispa**
## Current behavior before PR

Although more than one method can be assigned to an analysis and more than one instrument can be assigned to one method, is not possible to assign more than one method to the same instrument.

## Desired behavior after PR is merged
In the Instrument's edit view, the user can assign more than one method to that instrument through a multi-selection widget.

Some codes from #1871 also were used.

--
I confirm I have read the [Bika LIMS Developer Guidelines][1] and that I have
tested the PR thoroughly and coded it according to [PEP8][2] standards.

[1] https://github.com/bikalabs/bika.lims/wiki/Bika-LIMS-Developer-Guidelines
[2] https://www.python.org/dev/peps/pep-0008

### CODE REVIEW

File: bika/lims/browser/instrument.py [hunk 1]
```diff
@@ -620,24 +620,37 @@ def __init__(self, context, request):
...
         bsc = getToolByName(self, 'bika_setup_catalog')
-        instrument = bsc(portal_type='Instrument', UID=self.request.get("uid", '0'))
-        if instrument and len(instrument) == 1:
-            method = instrument[0].getObject().getMethod()
-            if method:
-                methoddict = {'uid': method.UID(),
-                              'title': method.Title()}
-        return json.dumps(methoddict)
+        results = bsc(portal_type='Instrument', UID=self.request.get("uid", '0'))
+        instrument = results[0] if results and len(results) == 1 else None
```
Comment (ramonski) [inline bika/lims/browser/instrument.py]: I have seen this pattern quite a lot in the code and find it kind of cryptic.

We should better use a more pythonic approach like this:
```
results = bsc(portal_type='Instrument', UID=self.request.get("uid", '0'))
if results:
    instrument = results[0]
```

Also this code swallows an important exception, when we get _more_ than one result for an instrument UID. Would this case happen, then this function would return nothing.

So if this is really needed, we should fail:
```
if len(results) > 1:
    raise RuntimeError("More than one instrument found for UID %s" % self.request.get("uid", '0'))
```

File: bika/lims/content/analysisservice.py [hunk 2]
```diff
@@ -1211,10 +1209,10 @@ def getAvailableMethods(self):
             # Add the methods from the instruments capable to perform
             # this analysis service
             for ins in self.getInstruments():
-                method = ins.getMethod()
-                if method and method.UID() not in muids:
-                    methods.append(method)
-                    muids.append(method.UID())
+                for method in ins.getMethods():
+                    if method and method.UID() not in muids:
```
Comment (ramonski) [inline bika/lims/content/analysisservice.py]: Maybe a dictionary would be better here, e,g, `{method.UID(): method}` and return dict.values() at the end.

File: bika/lims/content/instrument.py [hunk 3]
```diff
@@ -391,6 +404,17 @@ def getManufacturers(self):
         items.sort(lambda x, y: cmp(x[1], y[1]))
         return DisplayList(items)
 
+    from bika.lims import deprecated
```
Comment (ramonski) [inline bika/lims/content/instrument.py]: This should go to the top of the file

File: bika/lims/exportimport/setupdata/__init__.py [hunk 4]
```diff
@@ -772,7 +772,7 @@ def Import(self):
             obj.setInstrumentType(instrumenttype)
             obj.setManufacturer(manufacturer)
             obj.setSupplier(supplier)
-            obj.setMethod(method)
+            obj.setMethod([method])
```
Comment (ramonski) [inline bika/lims/exportimport/setupdata/__init__.py]: good catch

File: bika/lims/upgrade/to320.py [hunk 5]
```diff
@@ -186,15 +190,6 @@ def create_CAS_IdentifierType(portal):
...
     """
-    pc = getToolByName(portal, 'portal_catalog', None)
-    objs = pc(portal_type="Analyses",review_state="to_be_verified")
-    for obj_brain in objs:
-        obj = obj_brain.getObject()
-        old_field = obj.Schema().get("NumberOfVerifications", None)
-        if old_field:
-            new_value=''
-            for n in range(0,old_field):
-                new_value+='admin'
```
Comment (ramonski) [inline bika/lims/upgrade/to320.py]: This change seems to not belong into this PR

File: bika/lims/upgrade/to320.py [hunk 6]
```diff
@@ -205,21 +200,39 @@ def create_CAS_IdentifierType(portal):
...
     Getting all analyses with review_state in to_be_verified and
     adding "admin" as a verificator as many times as this analysis verified before.
     """
     pc = getToolByName(portal, 'portal_catalog', None)
-    objs = pc(portal_type="Analyses",review_state="to_be_verified")
+    objs = pc(portal_type="Analyses", review_state="to_be_verified")
     for obj_brain in objs:
         obj = obj_brain.getObject()
-        old_field = obj.Schema().get("NumberOfVerifications", None)
+        old_field = obj.Schema().get("NumberOfVerifications", None).get(obj)
```
Comment (ramonski) [inline bika/lims/upgrade/to320.py]: better use `obj.getField("NumberOfVerifications")`

File: bika/lims/browser/instrument.py (hunk 1 above)
Comment (rockfruit) [inline bika/lims/browser/instrument.py]: perhaps simply `assert len(results) < 2`

Comment (rockfruit) [inline bika/lims/browser/instrument.py]: Modified to use uid_catalog - when using uid_catalog everyone (including all Plone code I browsed) assumes that UIDs are unique if present.

File: bika/lims/content/analysisservice.py (hunk 2 above)
Comment (rockfruit) [inline bika/lims/content/analysisservice.py]: done

File: bika/lims/upgrade/to320.py (hunk 5 above)
Comment (rockfruit) [inline bika/lims/upgrade/to320.py]: meh, it's an upgrade step, and a good fix, I'll leave it

**Comment by rockfruit** [comment]
Bansai!

File: bika/lims/content/analysisservice.py [hunk 7]
```diff
@@ -1186,14 +1186,12 @@ def getMethod(self):
...
+        # is correct.
         method = None
-        if (self.getInstrumentEntryOfResults() == True):
-            method = self.getInstrument().getMethod() \
-                if (self.getInstrument() \
-                    and self.getInstrument().getMethod()) \
-                else None
-        else:
-            method = self.get_Method();
+        if (self.getInstrumentEntryOfResults() is True):
```
Comment (rockfruit) [inline bika/lims/content/analysisservice.py]: If `Allow manual entry of results` is on, then we can select a `DefaultMethod` (_Method) from methods selected manually.

If `Allow instrument entry of results` is on, then we can select a `DefaultMethod` from a group of methods which are provided by selected instruments.

Why then, does the line above only return the _Method field (`DefaultMethod`) if the `InstrumentEntry` field is True?

Comment (xispa) [inline bika/lims/content/analysisservice.py]: I see @rockfruit , I'd say is better to return the first method from manual entry if the instrument entry of results is off:

    method = None
    if self.getInstrumentEntryOfResults():
        method = self.get_Method()
    else:
        method = self.getMethods()
        method = method[0] if method else None
    return method

**Pull request merged**